File: health-clock-backend/app/schemas/document.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class DocumentBase(BaseModel):
    member_id: str
    file_name: str = Field(..., max_length=255)
    file_size: int = Field(..., ge=0)
    mime_type: str = Field(..., max_length=100)
    category: Literal["checkup_report", "examination_result", "outpatient_record", "lab_report", "prescription", "hospitalization_record", "other"]
    title: str | None = Field(None, max_length=200)
    document_date: datetime | None = None
    hospital_name: str | None = Field(None, max_length=200)


class DocumentCreate(DocumentBase):
    file_url: str
    storage_bucket: str = Field(..., max_length=100)
    storage_key: str = Field(..., max_length=500)
    ocr_text: str | None = None
    ai_summary: dict | None = None


class DocumentUpdate(BaseModel):
    title: str | None = Field(None, max_length=200)
    document_date: datetime | None = None
    hospital_name: str | None = Field(None, max_length=200)
    category: str | None = None
    ocr_text: str | None = None
    ai_summary: dict | None = None


class Document(DocumentBase):
    id: str
    file_url: str
    storage_bucket: str
    storage_key: str
    ocr_text: str | None = None
    ai_summary: dict | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class UploadSignatureRequest(BaseModel):
    member_id: str
    file_name: str = Field(..., max_length=255)
    file_size: int = Field(..., ge=0, le=10 * 1024 * 1024)  # 最大 10MB
    mime_type: str = Field(..., max_length=100)


class UploadSignatureResponse(BaseModel):
    upload_url: str
    object_key: str
    file_url: str
    expires_in: int = 3600  # 签名有效期（秒）


class OCRRequest(BaseModel):
    document_id: str


class OCRResponse(BaseModel):
    document_id: str
    ocr_text: str
    ai_summary: dict | None = None
    candidate_events: list[dict] | None = None

File: health-clock-app/lib/shared/models/document.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'document.freezed.dart';
part 'document.g.dart';

@freezed
class HealthDocument with _$HealthDocument {
  const factory HealthDocument({
    required String id,
    required String memberId,
    required String fileName,
    required int fileSize,
    required String mimeType,
    required String category,
    String? title,
    DateTime? documentDate,
    String? hospitalName,
    required String fileUrl,
    required String storageBucket,
    required String storageKey,
    String? ocrText,
    Map<String, dynamic>? aiSummary,
    required DateTime createdAt,
    required DateTime updatedAt,
    String? downloadUrl,
  }) = _HealthDocument;

  factory HealthDocument.fromJson(Map<String, dynamic> json) =>
      _$HealthDocumentFromJson(json);
}

@freezed
class UploadSignatureRequest with _$UploadSignatureRequest {
  const factory UploadSignatureRequest({
    required String memberId,
    required String fileName,
    required int fileSize,
    required String mimeType,
  }) = _UploadSignatureRequest;

  factory UploadSignatureRequest.fromJson(Map<String, dynamic> json) =>
      _$UploadSignatureRequestFromJson(json);
}

@freezed
class UploadSignature with _$UploadSignature {
  const factory UploadSignature({
    required String uploadUrl,
    required String objectKey,
    required String fileUrl,
    @Default(3600) int expiresIn,
  }) = _UploadSignature;

  factory UploadSignature.fromJson(Map<String, dynamic> json) =>
      _$UploadSignatureFromJson(json);
}

@freezed
class DocumentCreate with _$DocumentCreate {
  const factory DocumentCreate({
    required String memberId,
    required String fileName,
    required int fileSize,
    required String mimeType,
    required String category,
    String? title,
    DateTime? documentDate,
    String? hospitalName,
    required String fileUrl,
    required String storageBucket,
    required String storageKey,
  }) = _DocumentCreate;

  factory DocumentCreate.fromJson(Map<String, dynamic> json) =>
      _$DocumentCreateFromJson(json);
}
